### ISSUE TITLE
pnet  cls_loss的问题

### ISSUE THREAD

**Issue by guojiapeng00** [AITTSMD/MTCNN-Tensorflow#77]
代码中关于pnet的损失是:     loss = -tf.log(label_prob+1e-10)
跟论文中的cross_entropy   = -y*log(p)  + (1-y) * (1-log（p))  好像不一样啊？
当y=1时两者相同，但当y = 0时，就不同了

**Comment by turboLIU** [comment]
@guojiapeng00 应该是避免了log（0）的尴尬

**Comment by guojiapeng00** [comment]
刚明白cls_loss   是用softmax计算的cross_entropy     ，loss =  -y*log(p),

**Comment by OceanWong1991** [comment]
def cls_ohem(cls_prob, label):
    zeros = tf.zeros_like(label)
    #label=-1 --> label=0net_factory
      # pos sample,cls_label=1, 将其他label置为0
    **label_filter_invalid = tf.where(tf.less(label,0), zeros, label)**
    num_cls_prob = tf.size(cls_prob)
     # 由两维数据reshape 成一维数据，即长度变成原来2倍
    cls_prob_reshape = tf.reshape(cls_prob,[num_cls_prob,-1])
    label_int = tf.cast(label_filter_invalid,tf.int32)
   # 取出数据长度
    num_row = tf.to_int32(cls_prob.get_shape()[0])
    **# row = [0, 2, 4, 6, 8, 10, ...................]**
    row = tf.range(num_row)*2
    关键下面两行代码，由于输出cls_prob格式[non-face-prob, face-prob], 
     第一行表示positive simple 取face-prob, 非 positive simple取non-face-prob index,
     因此loss function (1-y) * (1-log（p)) 消除掉了(positive simple，non-face-prob； 非 positive simple的non-face-prob index没有取) 
    **indices_ = row + label_int
    label_prob = tf.squeeze(tf.gather(cls_prob_reshape, indices_))**
    loss = -tf.log(label_prob+1e-10)
    zeros = tf.zeros_like(label_prob, dtype=tf.float32)
    ones = tf.ones_like(label_prob,dtype=tf.float32)
    valid_inds = tf.where(label < zeros,zeros,ones)
    num_valid = tf.reduce_sum(valid_inds)
    keep_num = tf.cast(num_valid*num_keep_radio,dtype=tf.int32)
    #set 0 to invalid sample
    loss = loss * valid_inds
    loss,_ = tf.nn.top_k(loss, k=keep_num)
    return tf.reduce_mean(loss)

**Comment by Dugudaibo** [comment]
因此loss function (1-y) * (1-log（p)) 消除掉了
这句话不是很懂，困扰了我好久了，还希望你能在详细解释一下~
蛮烦了~
@OceanWong1991

**Comment by OceanWong1991** [comment]
@Dugudaibo 参考一下这篇笔记，可能对于概念不理解[https://www.evernote.com/shard/s489/sh/9ebcbf8d-f8e7-430d-a15d-3973bea8535d/5f60371b09f661863d46c335129ad3cb](url)
@turboLIU loss = -tf.log(label_prob+1e-10)，这一句 + 1e-10避免log(0)

**Comment by WayneMark** [comment]
@Dugudaibo  通俗的理解是：y为1时，1-y=0，(1-y)*log(1-p)=0，最后loss=-log(p)；同理y为0时，y*log(p)=0，最后loss=-log(1-p)；也就是“positive就取face-prob index，非 positive simple取non-face-prob index“，就像@OceanWong1991 说的那样 “消除掉了positive simple，non-face-prob； 非 positive simple的non-face-prob index” ，被消掉是因为它们为0

**Comment by sycophant-stone** [comment]
> 因此loss function (1-y) * (1-log（p)) 消除掉了
> 这句话不是很懂，困扰了我好久了，还希望你能在详细解释一下~
> 蛮烦了~
> @OceanWong1991

这个一定要结合后面看.
在loss=-tf.log(label_prob+1e-10)时候的
1.	Label_prob包含 “neg label”取neg的概率
2.	“pos label”取pos的概率.
3.	Loss的值”不恰当地”让neg的也计算成了pos的公式:
a)	  Y* = 0时,neg,采用loss=1-log(p), 但这里使用的loss=-log(p)
b)	 Y* = 1 时, pos,采用loss=-log(p). 这里是正确的.
   一定要结合下面的看.
4.	在第三步计算的loss tensor中.会让loss和valid_inds相乘. 
a)	 Valid_inds由0,1组成数组, 0表示neg,1表示pos.
b)	Loss和valid_inds做点乘会导致,前面错误的y*计算(本来应该loss=1-log(p)采用的loss=-log(p))被用0抹掉了.就剩下pos的计算loss了.

**Comment by tangchen2** [comment]
> > 因此loss function (1-y) * (1-log（p)) 消除掉了
> > 这句话不是很懂，困扰了我好久了，还希望你能在详细解释一下~
> > 蛮烦了~
> > @OceanWong1991
> 
> 这个一定要结合后面看.
> 在loss=-tf.log(label_prob+1e-10)时候的
> 
> 1. Label_prob包含 “neg label”取neg的概率
> 2. “pos label”取pos的概率.
> 3. Loss的值”不恰当地”让neg的也计算成了pos的公式:
>    a)	  Y* = 0时,neg,采用loss=1-log(p), 但这里使用的loss=-log(p)
>    b)	 Y* = 1 时, pos,采用loss=-log(p). 这里是正确的.
>    一定要结合下面的看.
> 4. 在第三步计算的loss tensor中.会让loss和valid_inds相乘.
>    a)	 Valid_inds由0,1组成数组, 0表示neg,1表示pos.
>    b)	Loss和valid_inds做点乘会导致,前面错误的y*计算(本来应该loss=1-log(p)采用的loss=-log(p))被用0抹掉了.就剩下pos的计算loss了.

valid_inds中pos和neg 都是1，part，landmark才被置零了，所以这个错误计算的loss还是存在的
所以我觉得这里loss计算还是有问题的
